src/module.c

```diff
@@ -1099,6 +1099,33 @@ void jl_invalidate_binding_refs(jl_globalref_t *ref, jl_binding_partition_t *inv
     JL_GC_POP();
 }
 
+// Called for all GlobalRefs found in lowered code. Adds backedges for cross-module
+// GlobalRefs.
+JL_DLLEXPORT void jl_maybe_add_binding_backedge(jl_globalref_t *gr, jl_module_t *defining_module, jl_value_t *edge)
+{
+    if (!edge)
+        return;
+    // N.B.: The logic for evaluating whether a backedge is required must
+    // match the invalidation logic.
+    if (gr->mod == defining_module) {
+        // No backedge required - invalidation will forward scan
+        return;
+    }
+    jl_binding_t *b = gr->binding;
+    if (!b)
+        b = jl_get_module_binding(gr->mod, gr->name, 1);
+    if (!b->backedges) {
+        b->backedges = jl_alloc_vec_any(0);
+        jl_gc_wb(b, b->backedges);
+    } else if (jl_array_len(b->backedges) > 0 &&
+               jl_array_ptr_ref(b->backedges, jl_array_len(b->backedges)-1) == edge) {
```

Comment:
IIUC, this should be scanning the whole list, not just checking the last one (particularly because of the thread-safety issues this currently incurs)?